ImmutableStore Class

edbo opened the pull request:
This is a pull request that will replace the createImmutableStore function with an ES6 style class that you can now inherit to create a store. See the updated readme for usage.

yahoocla commented:
Thank you for submitting this pull request, however I do not see a valid CLA on file for you. Before we can merge this request please visit https://yahoocla.herokuapp.com/ and agree to the terms. Thanks! :smile:

edbo commented:
This isn't ready to be merge yet but I've created the stub so we close the other issue I raised and provide more context.

Running grunt you'll see that some tests fail which is related to this issue:

https://github.com/yahoo/fluxible-immutable-utils/issues/7

To fix this I would suggest we either:

*  Delete createImmutableMixin.js and the ImmutableMixin.js in favour of using the createImmutableContainer.js (my preference)
*  Fix the test

yahoocla commented:
CLA is valid!

3den commented on README.md:
```diff
@@ -172,47 +178,39 @@ and creates defaults for the following [fluxible store](http://fluxible.io/api/s
-    _onNewFoo: function (payload) {
-        this.setState(this._state.setIn(['data', payload.id], payload.foo));
-    },
+ContentStore.storeName = 'ContentStore';
```
can this be inside the class?

edbo replied:
Unfortunately not as there isn't a way to do that ES6 that I can find in the docs. We could build something to automatically name the Store after the class name maybe but then it would only get set after the constructor is called.

3den commented on README.md:
````diff
@@ -172,47 +178,39 @@ and creates defaults for the following [fluxible store](http://fluxible.io/api/s
-    }
-});
-```
+export default ContentStore;
````
the name of the class was FooStore

3den commented on lib/ImmutableStore.js:
```diff
@@ -0,0 +1,49 @@
+import {BaseStore} from 'fluxible/addons';
+import Immutable from 'immutable';
+
+class ImmutableStore extends BaseStore {
```
`export default class ImmutableStore extends BaseStore {`

3den commented on lib/ImmutableStore.js:
```diff
@@ -0,0 +1,49 @@
+    }
+}
+
+export default ImmutableStore;
```
we can delete this and do the export on the same line as the class is defined

edbo commented:
@3den Thanks for the code review! I've taken care of the changes you requested. Do you have any input on my questions in my second comment?

akshayp commented:
@edbo There's still value to having the mixin so I would prefer we keep it around and actually fix the test issues. We've seen significant performance impact when wrapping everything in higher order components which doesn't happen when using the mixin. For the number of components we use the mixin this is a pretty severe cost in a large application. On the flip side I do see value in exporting what is currently in the mixin as a ES6 class that implements the shouldComponentUpdate so that components can get it via extension vs using it as a mixin

koulmomo commented:
to make sure we aren't spreading `FUD`. The performance issues relating to higher-order components applies mostly to nodejs server-side reactjs rendering (though it is logical to assume that it will also apply to client-side first time renders, though not as noticeable to the end user).

Subsequent client-side renders (especially where using immutablejs) are still relatively fast/performant when using higher-order components.

edbo commented:
I've rebased and updated this now. Seems to pass all the appropriate tests now.

Note I've setup a new source directory that lints using es6 as the environment and creates a file in lib. Idea being that all the files can be migrated into the source directory as es6 and comiled into the lib directory.

edbo commented:
@3den Any feedback on this?

I'd like it to be merged if it can so we don't have to track our fork in our projects, thanks!

3den commented on src/ImmutableStore.js:
```diff
@@ -0,0 +1,58 @@
+import {BaseStore} from 'fluxible/addons';
+import Immutable from 'immutable';
+
+export default class ImmutableStore extends BaseStore {
```
:heart_eyes:

3den commented:
@edbo it looks great thanks :+1:

3den commented:
can you also update the version on the package?

edbo commented:
OK I bumped it to 0.5.1 and put the reference to the createImmutableStore back in the index.js so it doesn't break anyone else if they pull this.

I'm guessing we can maybe deprecate the createImmutableStore but maybe do it with a deprecation message and drop it later.

dmhood commented on lib/index.js:
```diff
@@ -1,4 +1,3 @@
 module.exports = {
-    createImmutableMixin: require('./createImmutableMixin'),
-    createImmutableStore: require('./createImmutableStore')
```
we still need to keep this for backwards compatibility correct?

edbo commented:
Thanks, have fixed the removed import in lib/index.js

edbo commented:
Let me know if you guys need anything else regarding this.

3den commented:
It looks good to me.

@dmhood @akshayp can we get that merged?

dmhood commented:
:+1: thanks!
